### ISSUE TITLE
Using different set of extra features at training time in multitask

### ISSUE THREAD

**Issue by muammar** [chemprop/chemprop#198]
I am training a multitask model as [described here](https://chemprop.readthedocs.io/en/latest/tutorial.html#missing-target-values), and I wanted to ask if the following is possible or some suggestions on how to implement that (it is going to be faster that way than if I got myself through the source code):

1. Suppose I have two targets, `a` and `b`, in regression mode, and I train a multitask MPNN algorithm. 
2. For one specific input, say `x1`, the targets can be `a = 10` and `b = 0.1`.  
3. I want to add some external features to `x1` (one-hot encoded vector) to help the algorithm distinguish between the two targets. 

What I need at training time is to pass the features `[1, 0]` when training to predict `a` and `[0, 1]` when predicting `b`. With what is available right now in Chemprop, this seems not feasible at training time. 

If I put this on pseudocode, it would look something like this: 

```
for input in data: 
   features = somestuff(input)
   if target == A:
      features += [1, 0]
   if target == B:
       features += [0, 1]
   output = model(features)
   loss = yourfavoritelosshere
   loss.backward()
   etc
```

For prediction, it seems trivial, though. 

I would appreciate any input. 

Best,

**Comment by cjmcgill** [comment]
Hi @muammar, could you clarify some more what you want to do?

In multitask operations, usually the targets have molecular features that are relevant to each other, so you will be better off training them together and predicting them together. Why not predict both A and B simultaneously? If the targets have unrelated features then you would be better off setting up separate models altogether and not training them at once.

Are you doing this because you have data that are missing one target or the other? Because if that's the case then you can have those fields in the data file be blank and the model will ignore them (unless you are using sklearn operation).

**Comment by muammar** [comment]
> Hi @muammar, could you clarify some more what you want to do?
> 

Yes, sure, I will try my best.

> 
> 
> In multitask operations, usually the targets have molecular features that are relevant to each other, so you will be better off training them together and predicting them together. Why not predict both A and B simultaneously? If the targets have unrelated features then you would be better off setting up separate models altogether and not training them at once.
> 
> 

I do not understand what you mean by "the targets have molecular features that are relevant to each other." Could you elaborate more on that, please? If I got this right,  in my case, they seem relevant because the predictive power improves and the data comes from the same type of experiment. 

When training the MPNN algorithm in Chemprop with multitask, you are doing `hard parameter sharing`. This means that we optimize the same NN for predicting all tasks simultaneously (I think the output is a tensor). At least, that's what I remember after inspecting the source code some months ago (?). I am working on this specific data for which I noted that multitask training improves the model's predictive power. For instance, if I merge data from sets `A`, `C`, and `D` to train the algorithm, I get a better predictive regression power than doing models for each set separately. Thus, sharing the parameters seems to have a positive effect (empirically) on the final model I built here. However, if I add a new set `B`, which is nothing but the same `A` lab experiment with more exposure time (some pharma stuff), then the same SMILES string can have a target value equals 10 for the case of `A` and 0.001 for `B`. I think the algorithm needs some extra help here to extract representations more effectively. Something like: hey, if this SMILES is predicting `A`, I want you to concatenate the one-hot encoded vector `[1, 0]` to the `rdkit2d` features, and if it is predicting `B` then concatenate `[0, 1]` (the rest of tasks would have their respective extra vectors to keep dimensionality consistent).  In summary, I want to give this extra information as input together with `rdkit2d` features with multitask training/prediction. Is this possible with Chemprop? For prediction, I know how to do this, but for training I don't think is feasible. 

Note: the hypothesis above comes from the fact that if I predict over a test set of `A`, results make sense, but predictions on `B` are a total disaster. 


> 
> Are you doing this because you have data that are missing one target or the other? Because if that's the case then you can have those fields in the data file be blank, and the model will ignore them (unless you are using sklearn operation).
> 

Some targets are missing, and they are blank on my CSV file.

**Comment by cjmcgill** [comment]
By "relevant to each other" I mean that training them together helps each other. The things that the model learns from A helps B and vice versa.

There is target scaling within chemprop, so even though A and B are different magnitudes, they will be scaled to be the same magnitude within the model. Just in case the magnitude of the values are what you are concerned with. There are some secondary functions that are unscaled but not usually significant.

You can add extra features manually during training using the argument `--features_path <path>` and indicating a csv containing a table of the same length as your datapoints, with the first row being headers. If you add extra features, you need to specify features in both training and prediction. These features are concatenated to the feature vector after the MPNN and before the FFNN, the same place as the features generator.
|A feature|B feature|
|---|---|
|1|0|
|0|1|
|0|1|


However, if you provide both a features path and a features generator, the features generator will override the path. The easiest way to use both kinds of information is to use the script `save_features.py` in the scripts directory to generate a list of the output from the generator and then concatenate that output with whatever other features you want to add and introduce them together though a path.

Something else you may want to try. There's a relatively new feature in chemprop for applying different relative weights to different targets in multitask training. If you wanted B to have twice the weight in training than A, C, and D, you could give the argument `--target_weights 1 2 1 1`. You supply a list of floats in the argument with the same length as your number of tasks. This feature works properly but we haven't done much testing to say what systems and situations it leads to better outcomes in. But it could be useful to try it out.

Some further follow up on the behavior you are seeing (okay if you don't have testing for these). Does the performance on A degrade when you add in B? If you ran BCD instead of ABCD, would B improve? From the verbose logs of the runs, is the best epoch falling much earlier for the ABCD runs compared to the ACD runs? Are the magnitudes of C and D more in the range of A or B? What do you mean by making predicting A *or* B (if the model is trained on A and B then it will output A and B) ?

**Comment by muammar** [comment]
> By "relevant to each other" I mean that training them together helps each other. The things that the model learns from A helps B and vice versa.
> 

OK, now I understand. 

> There is target scaling within chemprop, so even though A and B are different magnitudes, they will be scaled to be the same magnitude within the model. Just in case the magnitude of the values are what you are concerned with. There are some secondary functions that are unscaled but not usually significant.
> 

I had completely forgotten about this. I remember I also saw the scaling for targets by default. 

> You can add extra features manually during training using the argument `--features_path <path>` and indicating a csv containing a table of the same length as your data points, with the first row being headers. If you add extra features, you need to specify features in both training and prediction. These features are concatenated to the feature vector after the MPNN and before the FFNN, the same place as the features generator.
> 
> A feature	B feature
> 1	0
> 0	1
> 0	1


I have done this before, but will it work in the following scenario? 

The data for training looks like this: 

| SMILES      | A | B|
| ----------- | ----------- |--------|
| CC      | 10.0       | 1. |
| CCC   | 20.0        |0.002|
| C   | 5.0        ||


Then I would want the extra features to look like this: 

| SMILES      | A | B|
| ----------- | ----------- |--------|
|C       | [1, 0]  | |
| CC      | [1, 0]       | [0, 1] |
| CCC   | [1, 0]        | [0, 1]  |
|CCCC       |   | [0, 1]|

I could concatenate them and have a single vector like: 

| SMILES      | features |
| ----------- | --------|
|C       | [1, 0, 0, 0] | |
| CC      | [1, 0, 0, 1] |
| CCC   | [1, 0, 0, 1]  |
| CCCC   | [0, 0, 0, 1]  |

Note: I add a padding 00 for the case where there is no target data. 

That will look in the CSV file like: 

|SMILES | a1| a2| b1| b2|
|--------|--|----|--|--|
|C|       1| 0| 0| 0|
|CC|     1| 0| 0| 1|
|CCC|  1| 0| 0| 1|
|CCCC|  0| 0| 0| 1|

Probably this will not have the effect I have in mind; who knows. 

> However, if you provide both a features path and a features generator, the features generator will override the path. The easiest way to use both kinds of information is to use the script `save_features.py` in the scripts directory to generate a list of the output from the generator and then concatenate that output with whatever other features you want to add and introduce them together though a path.
> 

This is correct. I have done it that way before. I saved the features, load them in memory,  concatenated new ones, and then dump them into a numpy file and proceeded with training. 

> Something else you may want to try. There's a relatively new feature in chemprop for applying different relative weights to different targets in multitask training. If you wanted B to have twice the weight in training than A, C, and D, you could give the argument `--target_weights 1 2 1 1`. You supply a list of floats in the argument with the same length as your number of tasks. This feature works properly but we haven't done much testing to say what systems and situations it leads to better outcomes in. But it could be useful to try it out.
> 

Sounds interesting; I will look into that. 

> Some further follow up on the behavior you are seeing (okay if you don't have testing for these). Does the performance on A degrade when you add in B? If you ran BCD instead of ABCD, would B improve? From the verbose logs of the runs, is the best epoch falling much earlier for the ABCD runs compared to the ACD runs? Are the magnitudes of C and D more in the range of A or B? 

When I add `B` the performance of `A` does not degrade nor improve. It stays around the same. I have not tried the `BCD` approach yet. I will check the magnitudes; that sounds like a good thing to do. 

>What do you mean by making predicting A _or_ B (if the model is trained on A and B then it will output A and B) ?

Sorry,  my English is not good sometimes :S. When I predict with this multitask model, I get all outputs of `A` and `B` simultaneously. If I evaluate metrics for a test set for `A` predictions, I get better predictive power than a model only trained on `A`. When I evaluate metrics for a test set of `B`, then the predictive power is bad, e.g., negative $Rˆ2$, large MAE, and RMSE.  I still need to understand more about the lab experiments where this data is coming from. 

Thanks for discussing this with me.
